File: dailyproblem/07_03_2025/05_03_2025.cpp
class Solution {
    public:
        vector<int> closestPrimes(int left, int right) {
            vector<int> primes;
            for (int i = left; i <= right; i++){
                if i%1 ==0 && i%2 == 1 && i%3 == 1 && i%i == 0 && i != 1 && i != 2 && i != 3{
                    primes.push_back(i);
                }
                else if (i == 2 || i == 3)
                {
                    primes.push_back(i);
                }
                
            }
            int min_diff = INT_MAX;
            vector<int> res;
            for (int i = 0; i < primes.size() - 1; i++){
                if (primes[i + 1] - primes[i] < min_diff){
                    min_diff = primes[i + 1] - primes[i];
                    res = {primes[i], primes[i + 1]};
                }
            }
        }
        return res;
    };

    // class Solution {
    //     public:
    //         vector<int> closestPrimes(int left, int right) {
    //             // Define an upper bound for primes (Sieve of Eratosthenes)
    //             const int MAX = 1e6 + 1;
    //             vector<bool> isPrime(MAX, true);
    //             isPrime[0] = isPrime[1] = false; // 0 and 1 are not prime
        
    //             // Optimized Sieve of Eratosthenes
    //             for (int i = 2; i * i < MAX; i++) {
    //                 if (isPrime[i]) {
    //                     for (int j = i * i; j < MAX; j += i) {
    //                         isPrime[j] = false;
    //                     }
    //                 }
    //             }
        
    //             // Find first prime within range [left, right]
    //             int prevPrime = -1, minDiff = INT_MAX;
    //             vector<int> result = {-1, -1};
        
    //             for (int i = left; i <= right; i++) {
    //                 if (isPrime[i]) {
    //                     if (prevPrime != -1) {
    //                         int diff = i - prevPrime;
    //                         if (diff < minDiff) {
    //                             minDiff = diff;
    //                             result = {prevPrime, i};
    //                         }
    //                         // **Optimization**: If the difference is 2, it's the smallest possible prime gap, so we can stop early.
    //                         if (diff == 2) return result;
    //                     }
    //                     prevPrime = i;
    //                 }
    //             }
        
    //             return result;
    //         }
    //     };
        

File: dailyproblem/07_03_2025/05_03_2025.py
from typing import List

class Solution:
    def closestPrimes(self, left: int, right: int) -> List[int]:
        # Define the upper bound for sieve
        MAX = 10**6 + 1
        is_prime = [True] * MAX
        is_prime[0] = is_prime[1] = False  # 0 and 1 are not prime
        
        # Optimized Sieve of Eratosthenes
        for i in range(2, int(MAX ** 0.5) + 1):
            if is_prime[i]:
                for j in range(i * i, MAX, i):
                    is_prime[j] = False

        # Find the closest primes in range [left, right]
        prev_prime = -1
        min_diff = float("inf")
        result = [-1, -1]

        for num in range(left, right + 1):
            if is_prime[num]:
                if prev_prime != -1:
                    diff = num - prev_prime
                    if diff < min_diff:
                        min_diff = diff
                        result = [prev_prime, num]
                    # **Optimization**: Break early if minimum prime gap (2) is found
                    if diff == 2:
                        return result
                prev_prime = num

        return result
